sourcegraph/cody
# feat(uninstall): Delete cached credentials during re-install (CODY-1043)

This is the VSCode half of [CODY-1043](https://linear.app/sourcegraph/issue/CODY-1043/bug-account-credentials-are-cached-between-installs).

Unfortunately, we don't have access to any VSCode APIs during the uninstall script; it is just a raw node process that VSCode kicks off the next time VSCode is started after an uninstall (could be an arbitrary time after uninstalling). So the way we collected telemetry was to create files in the deactivation subroutine (which is run anytime the extension is stopped, i.e. when it is deactivated, VSCode is closed, the extension is upgraded or uninstalled) that stored all the config needed to boot telemetry and then we read those resources in the uninstall script.

Here we do basically the opposite. The uninstall script is run only when the user explicitly uninstalls the extension and closes and re-opens VSCode (importantly it does not run when the extension is upgraded). We create a marker file in the uninstallation process which we can check when the extension is initializing. When present, we delete the file and perform any cleanup from the previous installation (currently just deleting auth credentials and previous endpoints). 

This also fixes some existing errors in the uninstall script (it isn't working today because the client capabilities aren't initialized), adds a check to the bundling step for the uninstaller that no vscode dependencies are transitively included (similar to the vscode-shim) and adds a re-install telemetry event.

## Test plan
Finally added an E2E test for the uninstall flow. It took forever, but seeing as how it was broken for gosh knows how long and we didn't know indicates that it was necessary.

To manually test it:
1. Build the VSIX bundle with `pnpm -C vscode _build:vsix_for_test`
2. Start up VSCode and uninstall the Cody extension (or use the alternate editor, for instance I use VSCode Insiders day to day so I do this testing in VSCode).
3. From the command palette run "Extension: Install from VSIX" and select `vscode/dist/cody.e2e.vsix`. Log into Cody normally.
4. Close and reopen code. Verify that your credentials are cached.
5. Uninstall Cody (click on the gear icon on the extension view and select uninstall).
6. Fully close Code (not just the open window but the whole process)
7. Re-open Code and re-install Cody from VSIX.
8. Verify that credentials are cleared and you are presented with a fresh login screen.

## Changelog
deletes auth credentials and endpoint history when the extension is reinstalled
